File: frontend/src/types/achievement.ts
export interface Achievement {
  id: string;
  game_id?: string;
  console_id?: string;
  title: string;
  description: string;
  icon_url?: string;
  points: number;
  rarity: 'common' | 'rare' | 'epic' | 'legendary';
  criteria: Record<string, any>;
  created_at: string;
}

export interface UserAchievement {
  id: string;
  user_id: string;
  achievement_id: string;
  unlocked_at: string;
}

export interface UserAchievementProgress {
  total_achievements: number;
  unlocked_achievements: number;
  total_points: number;
  earned_points: number;
  completion_percentage: number;
}

export interface AchievementWithStatus extends Achievement {
  unlocked: boolean;
  unlocked_at?: string;
}

File: backend/src/models/achievement.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Achievement {
    pub id: Uuid,
    pub game_id: Option<Uuid>,
    pub console_id: Option<String>,
    pub title: String,
    pub description: String,
    pub icon_url: Option<String>,
    pub points: i32,
    pub rarity: String,
    pub criteria: serde_json::Value,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct UserAchievement {
    pub id: Uuid,
    pub user_id: Uuid,
    pub achievement_id: Uuid,
    pub unlocked_at: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct UserAchievementProgress {
    pub total_achievements: i64,
    pub unlocked_achievements: i64,
    pub total_points: i32,
    pub earned_points: i32,
    pub completion_percentage: f32,
}

#[derive(Debug, Deserialize)]
pub struct UnlockAchievementRequest {
    pub achievement_id: Uuid,
    pub game_id: Option<Uuid>,
}

#[derive(Debug, Serialize)]
pub struct AchievementWithStatus {
    #[serde(flatten)]
    pub achievement: Achievement,
    pub unlocked: bool,
    pub unlocked_at: Option<DateTime<Utc>>,
}
